ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,1035 @@
+the traffic closer to the intended destination, using only local knowledge. In
+practice, only nodes that are directly connected to more than one peer will
+forward traffic on behalf of other nodes.
+
+Knowing the destination locator of the traffic in question, and the locators of
+all directly connected peers, a node can work out which egress peering will take
+the traffic closer to the intended destination.
+
+---
+
+## Identities
+
+In order for Yggdrasil to operate a zero-trust model, node identities are
+generated cryptographically. These same cryptographic identities are then used
+for signing and encryption operations.
+
+Although it may often be beneficial for these identities to be persistent, there
+is no hard requirement for this. A node can either load existing persistent
+identities or generate new identities at startup, depending upon whether
+permanence is required.
+
+### Keypairs
+
+An Yggdrasil node has two identities: an `ed25519` keypair (referred to as the
+"signing keys") and a `curve25519` keypair (referred to as the "encryption
+keys"). While `ed25519` keys can indeed be converted into `curve25519` keys,
+there is no requirement for the two keypairs to be related - using separately
+generated keypairs is a fully supported configuration.
+
+The signing keys are critical in participating in the global spanning tree, both
+for performing root elections and signing switch root updates, as well as to
+generate a Tree ID. The encryption keys are used to encrypt traffic between
+nodes, as well as to generate a Node ID.
+
+### Tree ID
+
+The Tree ID is a 64-byte identifier which is calculated by taking the SHA512 sum
+of the node's public signing key.
+
+### Node ID
+
+The Node ID is a 64-byte identifier which is calculated by taking the SHA512 sum
+of the node's public encryption key. The node's permanent address is derived
+from the Node ID.
+
+### Strength
+
+When calculating the "strength" of a given Tree ID or Node ID, the entire
+identifier is treated as a 512-bit integer. The "strongest" ID is the one with
+the "highest" value, e.g. the highest number of leading one bits set.
+
+---
+
+## Wire protocol
+
+The Yggdrasil protocol is a binary protocol. Every message has one or more
+fields, matching one of the below base types. The first bytes of every message
+represent a `varu64` (see below) which defines the message type code. The
+structure of the message following that message type code must comply with that
+message type as defined.
+
+The protocol does not otherwise frame fields - that is, there are no markers
+within a message which denote the start or end of a given field. It is the
+responsibility of an implementation to know the format for a defined message
+type and to be able to determine the end of a given field based on the
+characteristics of that field type. Further details on this are available in the
+base type definitions below.
```

Comment:
This is not a great design, because it prevents you from defining new packets in a backwards-compatible way.